tsl/src/nodes/skip_scan/exec.c

```diff
@@ -0,0 +1,316 @@
+
+	if (has_nulls_first(state))
+		skip_scan_switch_stage(state, SS_NULLS_FIRST);
+	else
+		skip_scan_switch_stage(state, SS_NOT_NULL);
+}
+
+static bool
+has_nulls_first(SkipScanState *state)
+{
+	return state->nulls_first;
+}
+
+static bool
+has_nulls_last(SkipScanState *state)
+{
+	return !has_nulls_first(state);
+}
+
+static void
+skip_scan_rescan_index(SkipScanState *state)
+{
+	index_rescan(*state->scan_desc,
+				 *state->scan_keys,
+				 *state->num_scan_keys,
+				 NULL /*orderbys*/,
+				 0 /*norderbys*/);
+}
+
+/*
+ * Update skip scankey flags according to stage
+ */
+static void
+skip_scan_switch_stage(SkipScanState *state, SkipScanStage new_stage)
+{
+	Assert(new_stage > state->stage);
+
+	switch (new_stage)
+	{
+		case SS_NOT_NULL:
+			state->skip_key->sk_flags = SK_ISNULL | SK_SEARCHNOTNULL;
+			break;
+
+		case SS_VALUES:
+			state->skip_key->sk_flags = 0;
+			break;
+
+		case SS_NULLS_LAST:
+		case SS_NULLS_FIRST:
+			state->skip_key->sk_flags = SK_ISNULL | SK_SEARCHNULL;
+			state->skip_key->sk_argument = 0;
+			break;
+
+		default:
+			break;
+	}
+
+	state->stage = new_stage;
+}
+
+static void
+skip_scan_update_key(SkipScanState *state, TupleTableSlot *slot)
+{
+	if (!state->prev_is_null && !state->distinct_by_val)
+		pfree(DatumGetPointer(state->prev_datum));
+
+	MemoryContext old_ctx = MemoryContextSwitchTo(state->ctx);
+	state->prev_datum = slot_getattr(slot, state->distinct_col_attnum, &state->prev_is_null);
+	if (state->prev_is_null)
+	{
+		state->skip_key->sk_flags = SK_ISNULL;
+		state->skip_key->sk_argument = 0;
+	}
+	else
+	{
+		state->prev_datum =
+			datumCopy(state->prev_datum, state->distinct_by_val, state->distinct_typ_len);
+		state->skip_key->sk_argument = state->prev_datum;
+	}
+
+	MemoryContextSwitchTo(old_ctx);
+
+	/* we need to do a rescan whenever we modify the ScanKey */
+	skip_scan_rescan_index(state);
+}
+
+static TupleTableSlot *
+skip_scan_exec(CustomScanState *node)
+{
+	SkipScanState *state = (SkipScanState *) node;
+	TupleTableSlot *result;
+
+	switch (state->stage)
+	{
+		case SS_NULLS_FIRST:
+			result = state->idx->ps.ExecProcNode(&state->idx->ps);
+
+			/*
+			 * if we found a NULL value we return it, otherwise
+			 * we restart the scan looking for non-NULL
+			 */
+			skip_scan_switch_stage(state, SS_NOT_NULL);
+			skip_scan_rescan_index(state);
+			if (!TupIsNull(result))
+				return result;
+			else
+				return skip_scan_exec(node);
+
+			break;
+
+		case SS_NOT_NULL:
+		case SS_VALUES:
+			result = state->idx->ps.ExecProcNode(&state->idx->ps);
+
+			if (!TupIsNull(result))
+			{
+				/*
+				 * if we found a tuple we update the skip scan key
+				 * and look for values greater than the value we just
+				 * found. If this is the first non-NULL value we
+				 * also switch stage to look for values greater than
+				 * that in subsequent calls.
+				 */
+				if (state->stage == SS_NOT_NULL)
+					skip_scan_switch_stage(state, SS_VALUES);
+
+				skip_scan_update_key(state, result);
```

Comment:
return result should be right after 230, not after else? If we don't have  a valid tuple, we should not return at 251 and just switch states.

Reply:
we need to return there even for SS_END but i can put return on line 231 and after 249 to make it easier to follow

Reply:
or actually what you suggest would work too with another round trip to skip_scan_exec
current code saves us 1 roundtrip for the SS_END case